### PR TITLE
Additional methods for Chunk

### PR DESCRIPTION
**Pull request by pk044**
#2271

### CODE REVIEW

**Comment by jdegoes** [comment]
@pk044 Wow, this looks great!

I see one failing test (flaky), or maybe bug in Chunk (I didn't check):

```
[info]   - indexWhere
[info]     Test failed after 4 iterations with input: ((Singleton(),Fun( -> true)),-10)
[info]     Original input before shrinking was: ((Singleton(yUhq),Fun(yUhq -> true)),-6)
[info]     0 did not satisfy equalTo(-10)
[info] Done

```

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 1]
```diff
@@ -117,6 +117,83 @@ sealed trait Chunk[+A] { self =>
     case _ => false
   }
 
+  /**
+   * Returns the element that satisfies the predicate.
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/Chunk.scala]:
```suggestion
   * Returns the first element that satisfies the predicate.
```

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 2]
```diff
@@ -117,6 +117,83 @@ sealed trait Chunk[+A] { self =>
...
+    if (isEmpty) None else Some(self(0))
+
+  /**
+   * Returns the last element of this chunk if it exists.
+   */
+  def lastOption: Option[A] =
+    if (isEmpty) None else Some(self(self.length - 1))
+
+  /**
+   * Returns the index for which the given predicate is satisfied.
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/Chunk.scala]:
```suggestion
   * Returns the first index for which the given predicate is satisfied.
```

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 3]
```diff
@@ -117,6 +117,83 @@ sealed trait Chunk[+A] { self =>
...
+   */
+  def indexWhere(f: A => Boolean): Option[Int] =
+    indexWhere(f, 0)
+
+  /**
+   * Returns the index for which the given predicate is satisfied after or at some given index.
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/Chunk.scala]:
```suggestion
   * Returns the first index for which the given predicate is satisfied after or at some given index.
```

**Comment by pk044** [comment]
> @pk044 Wow, this looks great!
> 
> I see one failing test (flaky), or maybe bug in Chunk (I didn't check):
> 
> ```
> [info]   - indexWhere
> [info]     Test failed after 4 iterations with input: ((Singleton(),Fun( -> true)),-10)
> [info]     Original input before shrinking was: ((Singleton(yUhq),Fun(yUhq -> true)),-6)
> [info]     0 did not satisfy equalTo(-10)
> [info] Done
> ```

interesting. thanks - I'll look into it :)

**Comment by adamgfraser** [comment]
Looks like the flakiness is only occurring on Scala 2.11 on the JVM. May be a version / platform specific issue.

**Comment by adamgfraser** [comment]
It is actually a bug in Scala:

```scala
val as = List("")
as.indexWhere(_ => true, -10) // should be 0 but -10 on Scala 2.11.12
```

https://scastie.scala-lang.org/u0a6DuJ8SVa4FHWOaTTsnQ

@pk044 You can use test aspects or the `TestPlatform` and `TestVersion` functionality to check the version / platform and prevent the test from running on Scala 2.11 JVM.

I will check whether the bug has already been reported and if not file a bug report with Scala team.

**Comment by adamgfraser** [comment]
Ticket opened at https://github.com/scala/bug/issues/11852.

**Comment by adamgfraser** [comment]
Actually I don't think we have quite the functionality to do this in ZIO Test today (we have functionality for Scala 2 versus Dotty but not for specific versions within Scala 2). What if for purposes of this PR we only test for non-negative starting indices and I can work to add something to ZIO Test to address this.

**Comment by pk044** [comment]
@adamgfraser sure :) 

just so we don't forget about this once we are able to exclude this test on 2.11: should I make a GitHub issue or a TODO comment to make the test stop using non-negative indices?

**Comment by pk044** [comment]
@adamgfraser sure :) 

just so we don't forget about this once we are able to exclude this test on 2.11: should I make a GitHub issue or a TODO comment to make the test stop using non-negative indices?

**Comment by adamgfraser** [comment]
Sure. I'm working on right now so should be able to take care of today and fixing this will be a nice test case for the functionality.

**Comment by adamgfraser** [comment]
@pk044 Can you merge upstream changes? I think you are working off a version of master where we disabled Dotty support to get the new environment encoding PR merged so the required CI task is not being triggered.

**Comment by pk044** [comment]
@adamgfraser done! it seems like the tests passed on Dotty

**Comment by adamgfraser** [comment]
@pk044 Great work on this!

**Pull request merged**